state/protocol/badger/mutator_test.go

```diff
@@ -610,246 +617,396 @@ func TestExtendEpochTransitionValid(t *testing.T) {
+		)
 
-		// add blocks containing receipts for block1 and block2 (necessary for
-		// sealing)
+		// add blocks containing receipts for block1 and block2 (necessary for sealing)
+		// block 1 receipt contains nextEpochSetup1
 		block1Receipt := unittest.ReceiptForBlockFixture(&block1)
+		block1Receipt.ExecutionResult.ServiceEvents = []flow.ServiceEvent{nextEpochSetup1.ServiceEvent()}
+
+		// add block 1 receipt to block 3 payload
 		block3 := unittest.BlockWithParentFixture(block1.Header)
 		block3.SetPayload(flow.Payload{
 			Receipts: []*flow.ExecutionReceipt{block1Receipt},
 		})
 		err = state.Extend(&block3)
-		require.Nil(t, err)
+		require.NoError(t, err)
 
+		// block 2 receipt contains nextEpochSetup2
 		block2Receipt := unittest.ReceiptForBlockFixture(&block2)
+		block2Receipt.ExecutionResult.ServiceEvents = []flow.ServiceEvent{nextEpochSetup2.ServiceEvent()}
+
+		// add block 2 receipt to block 4 payload
 		block4 := unittest.BlockWithParentFixture(block2.Header)
 		block4.SetPayload(flow.Payload{
 			Receipts: []*flow.ExecutionReceipt{block2Receipt},
 		})
 		err = state.Extend(&block4)
-		require.Nil(t, err)
+		require.NoError(t, err)
 
-		rootSetup := seal.ServiceEvents[0].Event.(*flow.EpochSetup)
+		// seal for block 1
+		seal1 := unittest.Seal.Fixture(unittest.Seal.WithResult(&block1Receipt.ExecutionResult))
+
+		// seal for block 2
+		seal2 := unittest.Seal.Fixture(unittest.Seal.WithResult(&block2Receipt.ExecutionResult))
+
+		// block 5 builds on block 3, contains seal for block 1
+		block5 := unittest.BlockWithParentFixture(block3.Header)
+		block5.SetPayload(flow.Payload{
+			Seals: []*flow.Seal{seal1},
+		})
+		err = state.Extend(&block5)
+		require.NoError(t, err)
+
+		// block 6 builds on block 4, contains seal for block 2
+		block6 := unittest.BlockWithParentFixture(block4.Header)
+		block6.SetPayload(flow.Payload{
+			Seals: []*flow.Seal{seal2},
+		})
+		err = state.Extend(&block6)
+		require.NoError(t, err)
+
+		// block 7 builds on block 5, contains QC for block 7
+		block7 := unittest.BlockWithParentFixture(block5.Header)
+		err = state.Extend(&block7)
+		require.NoError(t, err)
+
+		// block 8 builds on block 6, contains QC for block 6
+		block8 := unittest.BlockWithParentFixture(block6.Header)
+		err = state.Extend(&block8)
+		require.NoError(t, err)
+
+		// should be able query each epoch from the appropriate reference block
+		setup1FinalView, err := state.AtBlockID(block7.ID()).Epochs().Next().FinalView()
+		assert.NoError(t, err)
+		require.Equal(t, nextEpochSetup1.FinalView, setup1FinalView)
+
+		setup2FinalView, err := state.AtBlockID(block8.ID()).Epochs().Next().FinalView()
+		assert.NoError(t, err)
+		require.Equal(t, nextEpochSetup2.FinalView, setup2FinalView)
+	})
+}
+
+// we should be able to have conflicting forks with two duplicate instances of
+// the same service event for the same epoch
+//
+//         /--B1<--B3(R1)<--B5(S1)<--B7
+// ROOT <--+
+//         \--B2<--B4(R2)<--B6(S2)<--B8
```

Comment:
is the following case same as this one? Or has any test case covered it already?
```
ROOT <--B1 <- B2(R1) <-- B3(S1) <--B4
                            ^----- B5
```

This is to test we are able to process `S1` twice.

Reply:
`TestExtendDuplicateEpochEvents` will cover processing the same service event twice, since we create the same service event on two different forks, and handle service events on both forks.